Add support for scs-eureka

dyroberts opened the pull request:
- Extract autoconfig (moved to new scs autoconfig modules)
 - Add `CfEurekaClientProcessor`
connected to #66

spencergibb commented:
Porting auto config seems out of character for this project. Seems like it should just get properties into the correct for so auto config in some other project could take advantage. If some auto config that was previously in SCS connectors but wanted to move to pure OSS, seems like spring-cloud-netflix would be a better fit for that.

markpollack commented on java-cfenv-boot-pivotal-scs/src/main/java/io/pivotal/cfenv/boot/scs/config/CfSpringCloudConfigClientProcessor.java:
```diff
@@ -13,7 +13,7 @@
  * limitations under the License.
  */
-package io.pivotal.cfenv.boot.scs;
+package io.pivotal.cfenv.boot.scs.config;
```
While moving to a config package is nicer, it isn't the kind of change to make in a point release (which I am thinking where we will add this functionality).

dyroberts replied:
Ok, that makes sense. I'll move it all into one big happy package! :)

dyroberts commented:
@spencergibb I can see the logic in what you're saying. In this case, I was following the pattern set by the scs config-server support, which includes some auto-config. In fact, I think the main reason there is an scs-specific artifact, here in this project, is to provide this auto-config. Maybe you could log an issue to move all of this to spring-cloud-netflix?

markpollack commented:
> Porting auto config seems out of character for this project. Seems like it should just get properties into the correct for so auto config in some other project could take advantage. If some auto config that was previously in SCS connectors but wanted to move to pure OSS, seems like spring-cloud-netflix would be a better fit for that.

That does sound reasonable, but while `EurekaInstanceAutoConfiguration` will not kick-in unless it is an app running on CF, 

```@Configuration
@ConditionalOnProperty("eureka.client.serviceUrl.defaultZone")
@ConditionalOnExpression("'${vcap.application.uris[0]:}'!='' || '${cf.instance.ip:}'!=''")
public class EurekaInstanceAutoConfiguration {
```

The class `EurekaOAuth2AutoConfiguration` looks like it would need some additional conditionals to similarly restrict when it kicks in?

Also, there are a couple of existing autoconfig classes in the released code
```
org.springframework.boot.autoconfigure.EnableAutoConfiguration=\
io.pivotal.cfenv.boot.scs.VaultTokenRenewalAutoConfiguration,\
io.pivotal.cfenv.boot.scs.PlainTextConfigClientAutoConfiguration
``` 

Looking into it now, I can't actually figure out how `PlainTextConfigClientAutoConfiguration` is being used since it produced a bean of the interface type `PlainTextConfigClient` and that doesn't seem to be used anywhere.

markpollack commented:
There is also `SanitizingEurekaInstanceConfigBean` in this autoconfig which may be something only needed in CF environments?

spencergibb commented:
Didn't notice the config server stuff.

spencergibb commented:
Glancing at config stuff, I'd say the same thing. Why do auto config here?

markpollack commented:
Would be a good opportunity to fill in the class level Javadoc with some description....

markpollack commented:
I don't see what `PlainTextConfigClientAutoConfiguration` is actually doing anyway...... maybe it is cruft moved over?

dyroberts commented:
@markpollack @spencergibb Regarding the auto-configuration, scs has historically provided its own starter library, as well as connector. Maybe it makes more sense for that to continue, replacing the connector with java-cfenv?

dyroberts commented:
Looking at the `EurekaOAuth2AutoConfiguration` and `ConfigClientOAuth2BootstrapConfiguration`, it feels like the oauth2 portion of the auto-config could be moved to spring-cloud-netflix and spring-cloud-config respectively.

dyroberts commented:
@markpollack @spencergibb I've added a commit to this pr to start using the new version of spring-security, as defined by the spring-boot bom.

dyroberts commented:
@markpollack @spencergibb As discussed I've removed all autoconfig from this library and moved it to a new module that will be included, along with this library, in the scs starters.

markpollack commented:
merged in 738960b4224d59818faa1fab95ce10f5c8cb2e2d

xyloman commented:
@markpollack do you know when the 1.1.2.RELEASE will be cut that contains these updates?  I have had a few teams recently ask about the support and they need this capability in order to move off of spring cloud connectors.

If you were looking for testers before moving over we might be able to carve off some time to help test.

markpollack commented:
Dylan and I made a release the 1.1.2 release a few min ago, give it a whirl.

xyloman commented:
@markpollack we took these updates for a spin and with the removal of EurekaOAuth2AutoConfiguration and ConfigClientOAuth2BootstrapConfiguration how would these properties be leveraged that java-cfenv-boot-pivotal-scs sets?

dyroberts commented:
Hi @xyloman. Sorry for the confusion. The official spring-cloud-services client libraries are migrating to java-cfenv and the AutoConfig etc has been moved there where it ultimately belongs. There should be a new release out later this week that will depend on java-cfenv-boot-pivotal-scs:1.1.2.RELEASE. After that, you'll be able to follow the docs on connecting client apps (https://docs.pivotal.io/spring-cloud-services/3-1/common/client-dependencies.html). For now, you can continue to use the previous java-cfenv release.
